crates/runtime/src/libcalls.rs

```diff
@@ -434,50 +438,89 @@ unsafe fn externref_global_set(vmctx: *mut VMContext, index: u32, externref: *mu
-    Err(
-        anyhow::anyhow!("unimplemented: wasm atomics (fn memory_atomic_notify) unsupported",)
-            .into(),
-    )
+    let instance = (*vmctx).instance_mut();
+    let (_, spot) = validate_atomic_addr(instance, memory, addr_index, 4, 4)?;
+
+    if count == 0 {
+        return Ok(0);
+    }
+
+    let unparked_threads = spot.map_or(0, |spot| spot.unpark(addr_index, count));
+
+    Ok(unparked_threads)
 }
 
 // Implementation of `memory.atomic.wait32` for locally defined memories.
 unsafe fn memory_atomic_wait32(
     vmctx: *mut VMContext,
     memory_index: u32,
-    addr: u64,
-    _expected: u32,
-    _timeout: u64,
+    addr_index: u64,
+    expected: u32,
+    timeout: u64,
 ) -> Result<u32, TrapReason> {
+    let timeout = (timeout as i64 >= 0).then(|| Duration::from_nanos(timeout));
+
     let memory = MemoryIndex::from_u32(memory_index);
-    let instance = (*vmctx).instance();
-    validate_atomic_addr(instance, memory, addr, 4, 4)?;
-    Err(
-        anyhow::anyhow!("unimplemented: wasm atomics (fn memory_atomic_wait32) unsupported",)
-            .into(),
-    )
+    let instance = (*vmctx).instance_mut();
+    let (addr, spot) = validate_atomic_addr(instance, memory, addr_index, 4, 4)?;
+
+    // SAFETY: `addr` was validated by `validate_atomic_addr` above.
+    let atomic = unsafe { &*(addr as *const AtomicU32) };
+
+    // We want the sequential consistency of `SeqCst` to ensure that the `load` sees the value that the `notify` will/would see.
+    // All WASM atomic operations are also `SeqCst`.
+    let validate = || atomic.load(Ordering::SeqCst) == expected;
+
+    do_atomic_wait(addr_index, timeout, spot, validate)
 }
 
 // Implementation of `memory.atomic.wait64` for locally defined memories.
 unsafe fn memory_atomic_wait64(
     vmctx: *mut VMContext,
     memory_index: u32,
-    addr: u64,
-    _expected: u64,
-    _timeout: u64,
+    addr_index: u64,
+    expected: u64,
+    timeout: u64,
 ) -> Result<u32, TrapReason> {
+    let timeout = (timeout as i64 >= 0).then(|| Duration::from_nanos(timeout));
+
     let memory = MemoryIndex::from_u32(memory_index);
-    let instance = (*vmctx).instance();
-    validate_atomic_addr(instance, memory, addr, 8, 8)?;
-    Err(
-        anyhow::anyhow!("unimplemented: wasm atomics (fn memory_atomic_wait64) unsupported",)
-            .into(),
-    )
+    let instance = (*vmctx).instance_mut();
+    let (addr, spot) = validate_atomic_addr(instance, memory, addr_index, 8, 8)?;
+
+    // SAFETY: `addr` was validated by `validate_atomic_addr` above.
+    let atomic = unsafe { &*(addr as *const AtomicU64) };
+
+    // We want the sequential consistency of `SeqCst` to ensure that the `load` sees the value that the `notify` will/would see.
+    // All WASM atomic operations are also `SeqCst`.
+    let validate = || atomic.load(Ordering::SeqCst) == expected;
+
+    do_atomic_wait(addr_index, timeout, spot, validate)
+}
+
+#[inline]
```

Comment:
It should be ok to drop the `#[inline]` here and largely let LLVM figure it out.

Reply:
removed